### PR TITLE
feat: Add FlowBench kit — automated testing & benchmarking for Lamatic

### PR DESCRIPTION
**Pull request by bharath-avl**
## What is FlowBench?

FlowBench runs your Lamatic flow against a set of test cases before you merge a change, not after — closer to CI for a normal codebase than a monitoring tool. Give it a flow ID and a JSONL file of test cases, and it executes each one, scores the output using local sentence-embedding similarity (no external API calls), times each request, and compares the run against a saved baseline so you can see exactly what got better or worse.

## Why this isn't a duplicate of similar kits here

llm-silent-failure-detector watches a flow in production after it's shipped. FlowBench runs before you ship, against a fixed set of test cases you control.

live-api-debugger is for stepping through one request interactively. FlowBench runs a whole test suite at once, no manual steps.

## Testing this locally

1. cd kits/flowbench/apps && npm install
2. cp .env.example .env — fill in LAMATIC_API_URL, LAMATIC_API_KEY, LAMATIC_PROJECT_ID (Settings → API Keys in Lamatic Studio)
3. npm run dev, open localhost:3000
4. Enter a Flow ID and upload sample-tests/example.jsonl, click Run Benchmark
5. Report shows pass/fail, latency, similarity, and regressions vs. baseline

## A couple of things worth knowing

Similarity scoring measures semantic closeness, not just correctness — a correct but differently-phrased answer can score lower than expected. More on this in the README's Known Limitations section.

No database, no auth — baselines are flat JSON files in .flowbench/, kept intentionally simple.

One more thing: the PR template that auto-populated this box referenced an older config.json/bundles/templates structure that doesn't match how this repo actually works now (confirmed against the current CONTRIBUTING.md and the reference kits). I replaced it with this instead of trying to force-fit the old format.

- Added the FlowBench kit for pre-merge testing and benchmarking Lamatic flows.
- Added documentation and configuration:
  - `README.md`, `agent.md`, and default constitution.
  - Environment templates and `.gitignore` rules.
  - Lamatic kit configuration and sample JSONL test cases.
- Added the demo flow definition in `flows/flowbench-demo-flow.ts`:
  - GraphQL request trigger node accepts a user prompt and model selection.
  - Generative LLM node loads the system/user prompts and model configuration.
  - API response node maps the generated response for consumers.
  - Edges connect trigger → LLM → response, with a direct trigger response edge.
- Added the demo flow’s prompt and model configuration files.
- Added the Next.js FlowBench dashboard:
  - Run form validates a flow ID and JSONL upload.
  - Home page lists historical runs with pass rates and latency.
  - Report page displays case results, outputs, similarity scores, latency, regressions, and improvements.
  - Added reusable Button, Form, Input, Label, and class-merging utilities.
- Added benchmark execution services:
  - Typed Lamatic GraphQL client with environment validation, error handling, request timing, and payload safety checks.
  - Bounded-concurrency flow execution with per-case timeouts and isolated errors.
  - Offline sentence-embedding similarity scoring using `Xenova/all-MiniLM-L6-v2`.
  - Run metrics including cosine similarity, percentile calculation, and run IDs.
- Added baseline and run persistence using flat JSON files under `.flowbench/`.
- Added baseline comparison logic for new, removed, improved, regressed, and newly failing cases, including blocking regression detection.
- Added orchestration to execute benchmarks, save runs, update baselines only for successful non-regressing runs, and retrieve history/details.
- Added tests covering baseline comparisons, latency and similarity regressions, improvements, case additions/removals, and pass/fail changes.
- Added Next.js, TypeScript, Tailwind CSS v4, PostCSS, shadcn/ui, and test configuration.

### CODE REVIEW

**Comment by bharath-avl** [comment]
@coderabbitai review

File: kits/flowbench/apps/actions/orchestrate.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  run: RunResult;
+  diff: CompareResult;
+  baselineUpdated: boolean;
+  error?: string;
+}
+
+export async function executeBenchmark(
+  flowId: string,
+  testCases: TestCaseInput[]
+): Promise<OrchestrationResult> {
```
[the remark this answers is not in the document]
Comment (bharath-avl) [reply kits/flowbench/apps/actions/orchestrate.ts:19]: This is intentional, not an oversight — FlowBench's whole purpose is testing arbitrary flow IDs a user provides at runtime through the dashboard, not just wrapping one fixed flow the way most kits do. The kit still ships a real demo flow (flowbench-demo-flow.ts) and honors the standard envKey pattern as a fallback when no runtime flowId is given — see the comment block at the top of executeFlow in lamatic-client.ts. Happy to make this distinction more prominent in the README if it's unclear from the code alone.

**Comment by bharath-avl** [comment]
A few items from this review I'm deliberately deferring rather than fixing in this PR, with reasoning:

**react-hook-form + zod + shadcn/ui in RunForm.tsx** — agreed this is the documented convention, and I'm not doing it here because retrofitting a working, tested form this late in the review cycle risks breaking something that currently works correctly (upload, validation, loading states, keyboard accessibility) for a stylistic/convention gain rather than a correctness one. Happy to follow up with this as a separate PR if that's preferred over holding this one open longer.

**Tailwind CSS v3 → v4 migration** — same reasoning. Real guideline gap (package.json, postcss.config.mjs, and globals.css would all need coordinated changes), but a full migration this late carries real risk of breaking the UI for a convention-compliance gain, not a functional one.

Everything else flagged in this review — path traversal, GraphQL injection, keyboard accessibility, concurrency, per-case error isolation, threshold duplication, missing test coverage, the coerceToString serialization boundary, and the OrchestrationResult type honesty fix — has been addressed and pushed. Let me know if the two deferred items above should block merge; happy to tackle them if so.

**Comment by bharath-avl** [comment]
@coderabbitai review

**Comment by bharath-avl** [comment]
> A few items from this review I'm deliberately deferring rather than fixing in this PR, with reasoning:
> 
> **react-hook-form + zod + shadcn/ui in RunForm.tsx** — agreed this is the documented convention, and I'm not doing it here because retrofitting a working, tested form this late in the review cycle risks breaking something that currently works correctly (upload, validation, loading states, keyboard accessibility) for a stylistic/convention gain rather than a correctness one. Happy to follow up with this as a separate PR if that's preferred over holding this one open longer.
> 
> **Tailwind CSS v3 → v4 migration** — same reasoning. Real guideline gap (package.json, postcss.config.mjs, and globals.css would all need coordinated changes), but a full migration this late carries real risk of breaking the UI for a convention-compliance gain, not a functional one.
> 
> Everything else flagged in this review — path traversal, GraphQL injection, keyboard accessibility, concurrency, per-case error isolation, threshold duplication, missing test coverage, the coerceToString serialization boundary, and the OrchestrationResult type honesty fix — has been addressed and pushed. Let me know if the two deferred items above should block merge; happy to tackle them if so.

Pushed fixes for everything else flagged since — path traversal guard in storage.ts, GraphQL identifier validation in lamatic-client.ts, per-case error isolation in orchestrator.ts + runFlow.ts, keyboard-accessible dropzone in RunForm.tsx, unified regression thresholds between metrics.ts and compareBaseline.ts, and the two missing test cases for removedCases/improvements.

Still holding on the two deferred items above (react-hook-form/zod/shadcn refactor, Tailwind v4 migration) — let me know if either should block merge.

**Comment by akshatvirmani** [comment]
@bharath-avl LGTM!

can you resolve the coderabbit comments before we merge?

**Comment by bharath-avl** [comment]
@akshatvirmani thank you akshat. i will resolve them by eod.

**Comment by bharath-avl** [comment]
@coderabbitai review

**Comment by bharath-avl** [comment]
@akshatvirmani hey akshat resolved all coderabbit reviews and for 2 issues i replied with supporting message. you can run '/validate' when you are ready

**Comment by bharath-avl** [comment]
hey @akshatvirmani, thanks for merging can you tell me the process for the next steps in the challenge?
